### ISSUE TITLE
C language #define incorrectly interpreted

### ISSUE THREAD

**Issue by PhilG57** [CoatiSoftware/SourcetrailBugTracker#679]
Sourcetrail Version 2019.1.11 64bit database version 23 --

Sourcetrail considers #define statements in header (.h) files to be macros.  For example this --
   #define	HAVEFUNCTIONCODE			/* manageio.c */
is considered to be a macro when it really is not.  

This statement --
  #define _upnest (giNestCount++)	/* bump depth count for listbox message indent */
was correctly identified as a macro.

Many thanks again.

**Comment by mlangkabel** [comment]
Hi @PhilG57, thanks for bringing this up. In our implementation we just build on top of the Clang compiler frontend, which treats all of these `#define`s as "macros". 
So you say that `#define`s with parameters shall be called macros while `#define`s without parameters shall be called "defines"?
Can you point us to a resource that defines this convention so that we can argue that change?

**Comment by PhilG57** [comment]
Malte,
   Good morning and thank you for the response.
   I’m no expert on actual C standards but my usage of #define does seem reasonable.  I did manage to find the following at Wiki –
      https://en.wikipedia.org/wiki/C_preprocessor
   which does seem to support my usage.  
 
…from Wiki…
Whenever the identifier appears in the source code it is replaced with the replacement token list, which can be empty. For an identifier declared to be a function-like macro, it is only replaced when the following token is also a left parenthesis that begins the argument list of the macro invocation. The exact procedure followed for expansion of function-like macros with arguments is subtle. 
Object-like macros were conventionally used as part of good programming practice to create symbolic names for constants, e.g., 
#define PI 3.14159
instead of hard-coding <https://en.wikipedia.org/wiki/Hard_coding>  numbers throughout the code. An alternative in both C and C++, especially in situations in which a pointer to the number is required, is to apply the const qualifier to a global variable. This causes the value to be stored in memory, instead of being substituted by the preprocessor. 
An example of a function-like macro is: 
#define RADTODEG(x) ((x) * 57.29578)

… end Wiki

In any case, whatever you and your team decides is fine with me.  Many thanks for the consideration.  I am finding your program to be of great value to me and hope that your program continues to support the C language.  Personally, I’m not much of a fan of C++…


---------------
.phil groben.

**Comment by PhilG57** [comment]
Malte,

   I did remember a copy of “The C Programmer’s Handbook” on my bookshelf written by Morris Bolsky at Bell Labs in which #define as a token replacement and as a macro is discussed.  The first definition (token) is “#define identifier string” and the second definition (macro) is “#define identifier1(identifier2, …) string.  Under the second (macro) definition it says “To avoid expression errors, enclose a macro parameter in parenthesis everywhere it occurs in a macro definition”.

   Anyway, that’s it.  Hope that helps.

   Thanks again. 

 

---------------

.phil groben.

 

From: Phil Groben [mailto:phil.groben@att.net] 
Sent: Monday, March 11, 2019 9:02 AM
To: 'CoatiSoftware/SourcetrailBugTracker'; 'CoatiSoftware/SourcetrailBugTracker'
Cc: 'Mention'; Phil Groben
Subject: RE: [CoatiSoftware/SourcetrailBugTracker] C language #define incorrectly interpreted (#679)

 

Malte,

   Good morning and thank you for the response.

   I’m no expert on actual C standards but my usage of #define does seem reasonable.  I did manage to find the following at Wiki –

       <https://en.wikipedia.org/wiki/C_preprocessor> https://en.wikipedia.org/wiki/C_preprocessor

   which does seem to support my usage.  

 

…from Wiki…

Whenever the identifier appears in the source code it is replaced with the replacement token list, which can be empty. For an identifier declared to be a function-like macro, it is only replaced when the following token is also a left parenthesis that begins the argument list of the macro invocation. The exact procedure followed for expansion of function-like macros with arguments is subtle. 

Object-like macros were conventionally used as part of good programming practice to create symbolic names for constants, e.g., 

#define PI 3.14159

instead of hard-coding <https://en.wikipedia.org/wiki/Hard_coding>  numbers throughout the code. An alternative in both C and C++, especially in situations in which a pointer to the number is required, is to apply the const qualifier to a global variable. This causes the value to be stored in memory, instead of being substituted by the preprocessor. 

An example of a function-like macro is: 

#define RADTODEG(x) ((x) * 57.29578)

… end Wiki

In any case, whatever you and your team decides is fine with me.  Many thanks for the consideration.  I am finding your program to be of great value to me and hope that your program continues to support the C language.  Personally, I’m not much of a fan of C++…

---------------

.phil groben.

**Comment by mlangkabel** [comment]
We just discussed this topic internally and came to the conclusion that we need to revisit how Sourcetrail handles macros. This distinction between function-like-macros and object-like-macros has not come up until now, so thank you very much for the input.

Lets go one step further: Say, we would implement different node types for these two kinds of macros that look differently in the graph. Can you tell us why you would like to have this distinction and how it would be useful?

**Comment by PhilG57** [comment]
I recently returned to work on a project I had left some months ago.  I wanted to quickly re-acquaint myself with the project and was mighty glad to discover Sourcetrail to help with that effort.  While beginning to learn Sourcetrail as to its capabilities and suitability to my needs, I discovered the “Overview” window.  There I was amazed to learn I had 133 macros when in fact I knew there were only two.  Once I figured out that discrepancy, I began to check for others, of which none so far have been found.

 

Basically I would ask Sourcetrail to be slightly more specific.  Or, once again, to accept an “.ini” file in which I could inform Sourcetrail to consider “#define SOMETHING” as a simple definition and not a macro.   A simple definition is easy to check out by looking in the header files for what is ordinarily a simple substitution.  A macro could also be simple but could also be much more complex, generating a bunch of code with unknown results.  When looking at a project it would be helpful to know up front whether there were only two, not 133, macros which needed in-depth investigation.

 

Once again, many thanks.

 

---------------

.phil groben.
